### PR TITLE
Resolves #1307 - BucketListIsConsistentWithDatabase Invariant

### LINKED ISSUE #1307: Write invariant for validating the BucketList state against the SQL database
This would allow detection of a bad SQL state (admin with sticky fingers, bugs, etc).

We probably need a couple flavors of this:
* a lightweight (ie, checks only for counts)
* a "full" (that checks for actual content).

### PR DESCRIPTION
**Pull request by jonjove**
Introduces BucketListIsConsistentWithDatabase invariant which resolves #1307. This invariant implements the newly introduced checkOnBucketApply to ensure that the database matches the bucket, after it was applied, for all ledgers starting from the first ledger considered in the bucket. Merge after #1317

### CODE REVIEW

**Comment by MonsieurNicolas** [comment]
Commit names must be useful when running `git log` (only thing that can reference github issues etc is the PR itself)

File: src/bucket/Bucket.h [hunk 1]
```diff
@@ -42,7 +43,29 @@ class Bucket : public std::enable_shared_from_this<Bucket>,
   public:
     // Helper class that reads through the entries in a bucket, used internally
     // during merging.
-    class InputIterator;
+    class InputIterator
```
Comment (MonsieurNicolas) [inline src/bucket/Bucket.h]: couple things:
* in general it's better to have a separate commit for code moves like this ("moved Bucket::InputIterator definition from cpp to h file")
* in this case, I don't see the reason to move it to the .h file, it just pollutes the .h file (it's only used internally), if all you wanted was to stop having inline declarations, you can still achieve the same thing by having both the class definition and implementation in the same file

Comment (MonsieurNicolas) [reply src/bucket/Bucket.h]: nm, I didn't see the rest of the changes in the file - was confused by the comment that is no longer relevant.

Comment (MonsieurNicolas) [inline src/bucket/Bucket.h]: I would remove this part of the comment:
> , used internally during merging.

Comment (MonsieurNicolas) [reply src/bucket/Bucket.h]: now that you are making this available outside of this file, it's the occasion to remove the code in `BucketApplicator` that does the same thing from what I can tell

File: src/invariant/BucketListIsConsistentWithDatabase.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        std::shared_ptr<Bucket> bucket)
+{
+    Bucket::InputIterator iter(bucket);
+    if (!Bucket::InputIterator(bucket))
+    {
+        return {};
+    }
+
+    uint64_t nAccounts = 0, nTrustLines = 0, nOffers = 0, nData = 0;
+    uint32 modifiedSinceLedgerSeq = std::numeric_limits<uint32>::max();
```
Comment (MonsieurNicolas) [inline src/invariant/BucketListIsConsistentWithDatabase.cpp]: from what I can tell the logic around `modifiedSinceLedgerSeq` is not correct, the value of `modifiedSinceLedgerSeq` is defined by which bucket we're checking in the [context of a bucketlist](https://github.com/stellar/stellar-core/blob/master/src/bucket/BucketList.h#L106-L108)
With the expected range figured out, you can then:
* ensure that the entries in this bucket are in the expected range (this probably belongs in the regular apply code, not in an invariant)
* query the SQL database for entries that fall into the range (well, with the upper bound set to max int this code)

File: src/invariant/BucketListIsConsistentWithDatabase.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    for ( ; iter; ++iter)
+    {
+        auto& e = *iter;
```
Comment (MonsieurNicolas) [inline src/invariant/BucketListIsConsistentWithDatabase.cpp]: should be `auto const&`

File: src/invariant/BucketListIsConsistentWithDatabase.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+#include "invariant/Invariant.h"
+
+namespace stellar
+{
+
+class Application;
+class Database;
+class LedgerDelta;
+
+class BucketListIsConsistentWithDatabase : public Invariant
```
Comment (MonsieurNicolas) [inline src/invariant/BucketListIsConsistentWithDatabase.h]: Please add some comments to explain what this invariant does

File: src/invariant/BucketListIsConsistentWithDatabase.cpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            case TRUSTLINE:
+                ++nTrustLines;
+                break;
+            case OFFER:
+                ++nOffers;
+                break;
+            case DATA:
+                ++nData;
+                break;
+            }
```
Comment (MonsieurNicolas) [inline src/invariant/BucketListIsConsistentWithDatabase.cpp]: we typically have to add (to avoid warnings)
```
default:
     abort(false);
```

**Comment by MonsieurNicolas** [comment]
forgot something else: you need to add a few test cases

**Comment by MonsieurNicolas** [comment]
also you need to add an index to the tables that you are querying with `lastmodified`

**Comment by jonjove** [comment]
I have updated the branch for people to review but it is not currently ready for merge

File: src/bucket/BucketTests.cpp [hunk 6]
```diff
@@ -835,6 +835,64 @@ TEST_CASE("bucket persistence over app restart", "[bucket][bucketpersist]")
     }
 }
 
+TEST_CASE("bucket count", "[bucket][count]")
```
Comment (MonsieurNicolas) [inline src/bucket/BucketTests.cpp]: I wonder if we should make `kNumLevels` a config (we can still hardcode that the app is never started with something other than 10 outside of the `--test` code path in main), such that we can test the behavior up to kNumLevels. Right now the test cases cannot spill up to 10, and therefore we cannot observe the "steady state" behavior from test cases.

File: src/bucket/BucketTests.cpp [hunk 7]
```diff
@@ -835,6 +835,64 @@ TEST_CASE("bucket persistence over app restart", "[bucket][bucketpersist]")
...
+
+        for (uint32_t level = 0; level < BucketList::kNumLevels - 1; level++)
+        {
+            CLOG(INFO, "Bucket") << "Checking Curr[" << level << "]";
+            auto curr = bl.getLevel(level).getCurr();
+            std::set<uint32_t> currLedgers;
+            uint32_t currMin = std::numeric_limits<uint32_t>::max();
+            for (Bucket::InputIterator iter(curr); iter; ++iter)
+            {
+                auto lastModified = EntryFrame::FromXDR((*iter).liveEntry())->getLastModified();
```
Comment (MonsieurNicolas) [inline src/bucket/BucketTests.cpp]: you don't need the `FromXDR` here

Comment (MonsieurNicolas) [reply src/bucket/BucketTests.cpp]: (you can access the `lastModified` field directly)

File: src/invariant/InvariantTests.cpp [hunk 8]
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
...
+
+    Work::State onSuccess() override
+    {
+        if (!mAdded)
+        {
+            auto& db = mApp.getDatabase();
+            size_t count = 0;
+            {
+                auto prep = db.getPreparedStatement(
+                    "SELECT"
```
Comment (MonsieurNicolas) [inline src/invariant/InvariantTests.cpp]: I would rather not have raw SQL code in test cases like this, and instead have those statements in the respective class, for example, here you can use the `*Frame::countObjects` methods - in order to avoid duplication, maybe we just need to have those methods take a min and max, and this test can just have `min==max` ?

File: src/invariant/InvariantTests.cpp [hunk 9]
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
...
+    LedgerEntryType mType;
+    bool mModified;
+
+    void maybeModifyAccountEntry()
+    {
+        auto& db = mApp.getDatabase();
+        std::string account;
+        {
+            auto prep = db.getPreparedStatement(
+                "SELECT accountid FROM accounts"
```
Comment (MonsieurNicolas) [inline src/invariant/InvariantTests.cpp]: you don't actually need this, standard load + check that last modified is what you expect does the same without introducing new SQL

File: src/invariant/InvariantTests.cpp [hunk 10]
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
...
+    }
+
+    void maybeModifyTrustLineEntry()
+    {
+        auto& db = mApp.getDatabase();
+        std::string account, issuer, assetCode;
+        uint32_t assetType;
+        {
+            auto prep = db.getPreparedStatement(
+                "SELECT accountid, issuer, assetcode, assettype FROM trustlines"
```
Comment (MonsieurNicolas) [inline src/invariant/InvariantTests.cpp]: same thing here - use regular code to load from database

File: src/invariant/InvariantTests.cpp [hunk 11]
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
...
+            st.exchange(soci::use(mFromLedgerSeq));
+            st.define_and_bind();
+            if (!st.execute(true))
+            {
+                return;
+            }
+            assert(!st.fetch());
+        }
+
+        LedgerEntry entry;
```
Comment (MonsieurNicolas) [inline src/invariant/InvariantTests.cpp]: this code goes away if you use the regular load method, right?

File: src/bucket/BucketList.cpp [hunk 12]
```diff
@@ -171,20 +171,108 @@ BucketList::mask(uint32_t v, uint32_t m)
...
+{
+    assert(level < kNumLevels);
+    if (level == 0)
+    {
+        return 1 + ledger % 2;
+    }
+
+    auto const size = levelSize(level);
+    auto const half = levelHalf(level);
+    if (mask(ledger, half) != 0 && level != 10)
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/bucket/BucketList.cpp]: Fixed

File: src/bucket/BucketTests.cpp (hunk 7 above)
Comment (jonjove) [reply src/bucket/BucketTests.cpp]: Fixed

File: src/invariant/InvariantTests.cpp (hunk 11 above)
Comment (jonjove) [reply src/invariant/InvariantTests.cpp]: The same comments as above apply to replacing the above SQL query with the regular load method. If I had a key for an entry with the appropriate ledger number (either through the above SQL query or through maintaining that information internally) and used the key to invoke the regular load method, this code would be simplified significantly. It is not so easy unfortunately. TrustFrame::loadTrustLine takes an Asset as a parameter. But to construct the object of type Asset (defined in XDR), would require much of this exact code specifically the if (assetType == ...) part. In net, little can be gained in terms of removing this code by using the regular load functions.

File: src/invariant/BucketListIsConsistentWithDatabase.cpp [hunk 13]
```diff
@@ -38,21 +39,16 @@ BucketListIsConsistentWithDatabase::getName() const
...
-    }
-
     uint64_t nAccounts = 0, nTrustLines = 0, nOffers = 0, nData = 0;
-    for ( ; iter; ++iter)
+    for (Bucket::InputIterator iter(bucket); iter; ++iter)
     {
         auto const& e = *iter;
         if (e.type() == LIVEENTRY)
         {
+            assert(EntryFrame::FromXDR(e.liveEntry())->getLastModified() >= oldestLedger);
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/invariant/BucketListIsConsistentWithDatabase.cpp]: 1. Fixed
2. Good catch. This was originally written as an assert to facilitate debugging, but it really should be a part of the invariant. I can also extend this check to make sure the lastModifiedLedgerSeq is no bigger than the upper bound, if you think that is worthwhile?

File: src/historywork/ApplyBucketsWork.cpp [hunk 14]
```diff
@@ -22,43 +22,35 @@ namespace stellar
...
-    , mFirstVerified(firstVerified)
     , mApplying(false)
     , mLevel(BucketList::kNumLevels - 1)
 {
     // Consistency check: LCL should be in the _past_ from firstVerified,
     // since we're about to clobber a bunch of DB state with new buckets
     // held in firstVerified's state.
-    auto lcl = app.getLedgerManager().getLastClosedLedgerHeader();
-    if (firstVerified.header.ledgerSeq < lcl.header.ledgerSeq)
+    /*auto lcl = app.getLedgerManager().getLastClosedLedgerHeader();
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/historywork/ApplyBucketsWork.cpp]: Fixed. In my opinion, this check should actually be moved upstream to where ApplyBucketsWork is created in CatchupMinimal. Thoughts?

File: src/historywork/CatchupMinimalWork.cpp [hunk 15]
```diff
@@ -131,10 +131,10 @@ CatchupMinimalWork::onSuccess()
     // Phase 3: apply the buckets.
     if (!mApplyWork)
     {
-        CLOG(INFO, "History") << "Catchup MINIMAL applying buckets for state "
-                              << LedgerManager::ledgerAbbrev(mFirstVerified);
+        //CLOG(INFO, "History") << "Catchup MINIMAL applying buckets for state "
+        //                      << LedgerManager::ledgerAbbrev(mFirstVerified);
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/historywork/CatchupMinimalWork.cpp]: This logging should not be removed, since it includes an important state update for CatchupMinimal. Will uncomment.

File: src/historywork/ApplyBucketsWork.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/historywork/ApplyBucketsWork.cpp]: yes that sounds good - @vogel will probably want to incorporate this in his changeset when he rebases

File: src/invariant/InvariantTests.cpp [hunk 16]
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
...
+
+template <typename T>
+void
+bucketListConsistent(
+        const std::function<std::vector<uint32_t>(Database&,
+                                                  LedgerEntryType,
+                                                  uint32_t)>& f)
+{
+    auto seed = std::chrono::system_clock::now().time_since_epoch().count();
+    std::default_random_engine gen(seed);
```
Comment (MonsieurNicolas) [inline src/invariant/InvariantTests.cpp]: forgot to mention: do not mess with the seed in test cases to keep test runs deterministic. If you need more randomness, iterate more.

File: src/invariant/InvariantTests.cpp (hunk 9 above)
Comment (MonsieurNicolas) [reply src/invariant/InvariantTests.cpp]: what I am advocating for is to avoid having SQL or soci code outside of the few classes that deal with it - it's fine to have helper methods exposed for tests only

File: src/bucket/BucketTests.cpp [hunk 17]
```diff
@@ -835,6 +835,59 @@ TEST_CASE("bucket persistence over app restart", "[bucket][bucketpersist]")
     }
 }
 
+TEST_CASE("bucket count", "[bucket][count]")
```
Comment (vogel) [inline src/bucket/BucketTests.cpp]: You are testing sizeOfCurr new methods of BucketList.cpp here, what do you think about adding tests just for them? Make a bunch of tuples {ledger, level, expectedSizeOfCurr, expectedSizeOfSnap, expectedOldestLedgerInCurr, expectedOldestLedgerInSnap} and so on?

File: src/bucket/BucketTests.cpp [hunk 18]
```diff
@@ -835,6 +835,59 @@ TEST_CASE("bucket persistence over app restart", "[bucket][bucketpersist]")
...
+{
+    VirtualClock clock;
+    Config cfg(getTestConfig());
+    Application::pointer app = Application::create(clock, cfg);
+    BucketList& bl = app->getBucketManager().getBucketList();
+    std::vector<LedgerKey> emptySet;
+
+    for (uint32_t nLedgers = 2; nLedgers <= 256; ++nLedgers)
```
Comment (vogel) [inline src/bucket/BucketTests.cpp]: isn't that more like currentLedger instead of "number of ledgers"?

File: src/bucket/BucketTests.cpp [hunk 19]
```diff
@@ -835,6 +835,59 @@ TEST_CASE("bucket persistence over app restart", "[bucket][bucketpersist]")
...
+                currLedgers.insert(lastModified);
+                currMin = std::min(currMin, lastModified);
+            }
+            REQUIRE(currLedgers.size() == BucketList::sizeOfCurr(nLedgers, level));
+            if (currMin != std::numeric_limits<uint32_t>::max())
+            {
+                REQUIRE(currMin == BucketList::oldestLedgerInCurr(nLedgers, level));
+            }
+
+            CLOG(INFO, "Bucket") << "Checking Snap[" << level << "]";
```
Comment (vogel) [inline src/bucket/BucketTests.cpp]: looks like exact copy of code above - candidate for function/lamda?

File: src/historywork/ApplyBucketsWork.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (vogel) [reply src/historywork/ApplyBucketsWork.cpp]: Good idea. We can also always leave assert() here and precondition comment to be extra safe.

File: src/historywork/ApplyBucketsWork.cpp [hunk 20]
```diff
@@ -20,51 +22,43 @@ namespace stellar
...
 }
 
-std::shared_ptr<Bucket>
+std::shared_ptr<Bucket const>
 ApplyBucketsWork::getBucket(std::string const& hash)
 {
-    std::shared_ptr<Bucket> b;
-    if (hash.find_first_not_of('0') == std::string::npos)
+    std::shared_ptr<Bucket const> b;
+    if (isZero(hexToBin256(hash)))
```
Comment (vogel) [inline src/historywork/ApplyBucketsWork.cpp]: I like that!

File: src/invariant/BucketListIsConsistentWithDatabase.cpp [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+    return "BucketListIsConsistentWithDatabase";
+}
+
+std::string
+BucketListIsConsistentWithDatabase::checkOnBucketApply(
+        std::shared_ptr<Bucket const> bucket,
+        uint32_t oldestLedger)
+{
+    uint64_t nAccounts = 0, nTrustLines = 0, nOffers = 0, nData = 0;
+    for (Bucket::InputIterator iter(bucket); iter; ++iter)
```
Comment (vogel) [inline src/invariant/BucketListIsConsistentWithDatabase.cpp]: I think we could add support for for-range for Bucket::InputIterator, should not be hard, but is it worth the time?

File: src/invariant/BucketListIsConsistentWithDatabase.cpp [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        auto const& e = *iter;
+        if (e.type() == LIVEENTRY)
```
Comment (vogel) [inline src/invariant/BucketListIsConsistentWithDatabase.cpp]: why are you not testing DEADENTRY here? is that impossible (I'm not very familiar with Buckets code, just curious)

File: src/invariant/InvariantTests.cpp [hunk 23]
```diff
@@ -122,3 +621,82 @@ TEST_CASE("onLedgerClose fail/succeed", "[invariant]")
             app->getInvariantManager().checkOnLedgerClose(tsfp, ld));
     }
 }
+
+TEST_CASE("onBucketApply fail/succeed", "[invariant]")
```
Comment (vogel) [inline src/invariant/InvariantTests.cpp]: I would prefer to have these tests in separate file - BucketListIsConsistentWithDatabaseTests

File: src/invariant/InvariantTests.cpp [hunk 24]
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
...
+                auto ef = EntryFrame::FromXDR(entry);
+                ef->storeAdd(ld, db);
+                mAdded = true;
+            }
+        }
+        return ApplyBucketsWork::onSuccess();
+    }
+};
+
+class ApplyBucketsWorkDeleteEntry : public ApplyBucketsWork
```
Comment (vogel) [inline src/invariant/InvariantTests.cpp]: I think you could also consider making one class that accepts lambda executed once on onSuccess here instead of 3 classes.

File: src/invariant/InvariantTests.cpp [hunk 25]
```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
...
+        return "accountdata";
+    default:
+        REQUIRE(false);
+        return {};
+    }
+}
+
+template <typename T>
+void
+bucketListConsistent(
```
Comment (vogel) [inline src/invariant/InvariantTests.cpp]: please find a better name for that function

File: src/bucket/BucketTests.cpp (hunk 17 above)
Comment (jonjove) [reply src/bucket/BucketTests.cpp]: I can definitely try to do this, but those functions are quite complicated and defined recursively so I would have to figure out how to evaluate them without using the implementations. Perhaps an easier idea is to try to write a test case which validates some invariants among those formulae, such as `oldestLedgerInSnap(ledger, level) + sizeOfSnap(ledger, level) == oldestLedgerInCurr(ledger, level)`. I could also add tests like `sizeOfSnap(ledger, level) == levelHalf(level)` implies `sizeOfSnap(ledger2, level) == levelHalf(level)` for `ledger2 > ledger`. Let me know if you think these kinds of tests are sufficient or if you would prefer me try to do some direct calculations.

Comment (vogel) [reply src/bucket/BucketTests.cpp]: What about invariants and some edge cases?

File: src/bucket/BucketTests.cpp (hunk 6 above)
Comment (jonjove) [reply src/bucket/BucketTests.cpp]: In order to create a proper test case for issue #1300, we will need to be able to get entries into the bottom level of the BucketList since it is the only place where a merge can cause the elimination of a DEADENTRY and corresponding LIVEENTRY.

File: src/bucket/BucketTests.cpp [hunk 26]
```diff
@@ -81,6 +81,57 @@ countEntries(std::shared_ptr<Bucket> bucket)
     auto pair = bucket->countLiveAndDeadEntries();
     return pair.first + pair.second;
 }
+
+void
+checkBucketCount(BucketList& bl, uint32_t ledgerSeq, uint32_t level, bool isCurr)
+{
+    auto bucket = bl.getLevel(level).getCurr();
+    auto sizeOf = BucketList::sizeOfCurr;
```
Comment (MonsieurNicolas) [inline src/bucket/BucketTests.cpp]: rename `sizeOf` -> `sizeOfBucket`: it's too close to `sizeof`

File: src/bucket/BucketTests.cpp [hunk 27]
```diff
@@ -81,6 +81,57 @@ countEntries(std::shared_ptr<Bucket> bucket)
     auto pair = bucket->countLiveAndDeadEntries();
     return pair.first + pair.second;
 }
+
+void
+checkBucketCount(BucketList& bl, uint32_t ledgerSeq, uint32_t level, bool isCurr)
```
Comment (MonsieurNicolas) [inline src/bucket/BucketTests.cpp]: I don't think this function actually counts anything: it's checking if the bucket's lower bound matches its ledger entries.
Also, can't we also check the upper bound at the same time?

File: src/invariant/BucketListIsConsistentWithDatabase.cpp [hunk 28]
```diff
@@ -0,0 +1,134 @@
...
+std::string
+BucketListIsConsistentWithDatabase::checkOnBucketApply(
+        std::shared_ptr<Bucket const> bucket,
+        uint32_t oldestLedger,
+        uint32_t newestLedger)
+{
+    uint64_t nAccounts = 0, nTrustLines = 0, nOffers = 0, nData = 0;
+    for (Bucket::InputIterator iter(bucket); iter; ++iter)
+    {
+        auto const& e = *iter;
```
Comment (MonsieurNicolas) [inline src/invariant/BucketListIsConsistentWithDatabase.cpp]: to make this code bullet proof, we should also check that 2 consecutive entries are strictly ordered

Comment (jonjove) [reply src/invariant/BucketListIsConsistentWithDatabase.cpp]: Done

File: src/bucket/BucketTests.cpp (hunk 26 above)
Comment (jonjove) [reply src/bucket/BucketTests.cpp]: Done

File: src/bucket/BucketTests.cpp [hunk 29]
```diff
@@ -81,6 +81,68 @@ countEntries(std::shared_ptr<Bucket> bucket)
...
+    if (ubound > 0)
+    {
+        REQUIRE(ubound == oldestLedger + sizeOfBucket - 1);
+    }
+}
+
+// If pred is false for ledger < L and true for ledger >= L then
+// binarySearchForLedger will return L.
+uint32_t
+binarySearchForLedger(
```
Comment (vogel) [inline src/bucket/BucketTests.cpp]: strange that there is no std algorithm for that...

File: src/historywork/ApplyBucketsWork.cpp [hunk 30]
```diff
@@ -99,9 +85,18 @@ void
 ApplyBucketsWork::onStart()
 {
     auto& level = getBucketLevel(mLevel);
-    HistoryStateBucket& i = mApplyState.currentBuckets.at(mLevel);
+    HistoryStateBucket const& i = mApplyState.currentBuckets.at(mLevel);
     if (mApplying || i.snap != binToHex(level.getSnap()->getHash()))
     {
+        if (!mApplying)
```
Comment (vogel) [inline src/historywork/ApplyBucketsWork.cpp]: this will only be called when `i.snap != binToHex(level.getSnap()->getHash()` is true
what about moving that out of outside if so it becames
```
auto isCurrentSnapshot = i.snap != binToHex(level.getSnap()->getHash();
if (!isCurrentSnapshot) { ...}
if (mApplying || !isCurrentSnapshot) {...}
```

File: src/historywork/ApplyBucketsWork.cpp [hunk 31]
```diff
@@ -111,6 +106,15 @@ ApplyBucketsWork::onStart()
     }
     if (mApplying || i.curr != binToHex(level.getCurr()->getHash()))
     {
+        if (!mApplying)
+        {
+            uint32_t oldestLedger =
```
Comment (vogel) [inline src/historywork/ApplyBucketsWork.cpp]: this looks exactly like code above, maybe a method?

File: src/ledger/AccountFrame.cpp [hunk 32]
```diff
@@ -353,6 +356,47 @@ AccountFrame::countObjects(soci::session& sess)
     return count;
 }
 
+uint64_t
+AccountFrame::countObjects(soci::session& sess,
```
Comment (vogel) [inline src/ledger/AccountFrame.cpp]: I think that you could use LedgerRange class here (from catchup-at PR, now in master).
countObjects(sess, int, int) looks cryptic from API point of view
countObjectsInRange(sess, int, int) is better
but
countObjects(sess, LedgerRange) is best ;)

File: src/ledger/AccountFrame.cpp [hunk 33]
```diff
@@ -353,6 +356,47 @@ AccountFrame::countObjects(soci::session& sess)
...
+{
+    uint64_t count = 0;
+    sess << "SELECT COUNT(*) FROM accounts"
+            " WHERE lastmodified >= :v1 AND lastmodified <= :v2;",
+         into(count), use(oldestLedger), use(newestLedger);
+    return count;
+}
+
+void
+AccountFrame::deleteAccounts(Database& db, uint32_t oldestLedger)
```
Comment (vogel) [inline src/ledger/AccountFrame.cpp]: By the name and argument types I would expect that to remove accounts on specific ledger, not from to.
I think it needs better name, like deleteAccountModifiedOnOrAfter (seems long) or deleteAccountFrom (not as clear).

Do you have any better idea?

**Comment by MonsieurNicolas** [comment]
@jonjove it looks like this PR needs to be rebased again

**Comment by jonjove** [comment]
Rebased, cherry-picked from #1364, and changed `Application::create` to `createTestApplication` in BucketListIsConsistentWithDatabaseTests.cpp, InvariantTests.cpp, and BucketTests.cpp

**Comment by MonsieurNicolas** [comment]
r+ 811b01c67e8bf10e7fa2fd739a24d8b97a540cbc

**Comment by MonsieurNicolas** [comment]
github sorts by date not by commit... trying again

**Comment by MonsieurNicolas** [comment]
r+ 4df5454b75ff2326288d0d50c02e93852b408aa3

**Comment by jonjove** [comment]
Fixed compile error

**Comment by MonsieurNicolas** [comment]
r+ 35dbc2cae4b0d7ef9cd4325bd93073d5edfa7e1f

**Comment by MonsieurNicolas** [comment]
r+ 5bedc6771bea9c146a3bbfa55e2dbeecedd263e4
